Add generalized harmonic upwind flux

geoffrey4444 opened the pull request:
Depends on #1496 

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -129,4 +132,135 @@ struct ComputeNormalDotFluxes {
+ * \ingroup NumericalFluxesGroup
+ * \brief Computes the generalized-harmonic upwind flux
+ *
+ * The upwind flux in general is given by Eq. (6.3) of \cite Teukolsky2015ega:
```
I think you need a space before the `:`

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -129,4 +132,135 @@ struct ComputeNormalDotFluxes {
+  // `dg::SendBoundaryFluxes` calls `package_data` to store these tags in a
+  // Variables. Local and remote values of this data are then combined in the
+  // `()` operator.
+
```
Remove line

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -8,9 +8,31 @@
 // IWYU pragma: no_forward_declare Tensor
+namespace {
+template <typename FieldTag>
+typename FieldTag::type weight_char_field(
```
Should we always use `db::item_type` instead of `::type`? Perhaps @nilsdeppe can you answer this

geoffrey4444 replied:
I'll leave this as-is unless there's a consensus that using ::type is problematic

nilsvu replied:
`::type` doesn’t work for compute tags. I think for DataBox tags we should always use `db::item_type`, if only to save someone who copies this code some trouble.

geoffrey4444 replied:
OK, whenever I refer to tag types, I'll use db::item_type

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +329,108 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  get<::Tags::CharSpeed<Tags::UMinus<Dim, Frame::Inertial>>>(*packaged_data) =
+      char_speed_u_minus;
+  get<Tags::ConstraintGamma2>(*packaged_data) = gamma2;
+  get<Tags::UnitNormalOneForm<Dim, Frame::Inertial>>(*packaged_data) =
```
In the other fluxes we contract with the unit normal here in `package_data` already, then store the result in a bunch of tags that are private to the flux struct and retrieve them in `operator()`. Can you also do this here?

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -117,5 +117,13 @@ struct EvolvedFieldsFromCharacteristicFields : db::SimpleTag {
 };
+
+template <size_t Dim, typename Frame>
+struct UnitNormalOneForm : db::SimpleTag {
```
If you really end up needing this (see comment above), this should be private to the flux struct. We have a `Normalized<UnnormalizedFaceNormal>` Tag for general use.

geoffrey4444 commented:
After discussing with @teukolsky, I modified this pull request to *not* average when computing the upwind flux. The *hpp file documents exactly what the current version does. Briefly, here's what it does:

1. Multiply the positive char speeds (computed using only interior quantities) by the corresponding interior char fields
2. Multiply the negative char speeds (computed using only interior quantities) by the corresponding exterior char fields
3. Add the result from steps one and 2
4. Transform the result from step 3 to evolved variables, using gamma2, the normal from the interior, and the result of step 3.

I think I still don't have the algorithm right, though. I would have expected I'd get the same result after step 4 (up to a sign) if I repeat the steps swapping interior <-> exterior quantities, but I don't. (Should I?)

@teukolsky @fmahebert do you have any ideas of how I can fix my algorithm?

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +130,135 @@ struct ComputeNormalDotFluxes {
+  // `package_data` to provide the data needed to compute the numerical fluxes.
+  using argument_tags = tmpl::list<
+      Tags::UPsi<Dim, Frame::Inertial>, Tags::UZero<Dim, Frame::Inertial>,
+      Tags::UPlus<Dim, Frame::Inertial>, Tags::UMinus<Dim, Frame::Inertial>,
```
Something is confusing me here with the $u^+$ and $u^-$. This is probably because I don’t understand the formulation well enough. If the - and + only mean „internal“ and „external“ to an element interface (as you describe in the documentation above), then these shouldn’t be both argument/package tags. The packaging is done on both sides separately, then communicated and combined in the `operator()`. So I think the tags should be more like the following:
- The argument tags should be evolved or char fields, and the char speeds
- The packaged tags should either be those again, or even better as far as you can compute with the data that’s available on that side of the interface alone (e.g. $S \Lambda S^-1 u$)
- The `operator()` then takes the packaged data from both sides of the interface as arguments (as the comments in the code, presumably copied from the elliptic code, describe). So they will be $u^-$ and $u^+$ automatically (meaning internal and external to the element interface). 

My confusion is probably due to the same +- notation for the characteristic variables, and I‘m having trouble distinguishing them.

**Edit:** In terms of what you wrote above, I think you should do steps 1. and 2. in `package_data`, perhaps even step 4. if that‘s possible, and do step 3. in `operator()`.

geoffrey4444 replied:
Thanks for your comments, @nilsleiffischer! In Generalized harmonic, the char fields are named U^+, U^-, U^0, and U^\Psi. This use of the plus/minus superscript does not refer to evaluating on one side of the interface or the other.

nilsvu commented:
@geoffrey4444 Didn’t see your comment until now. Perhaps what I wrote on UPlus and UMinus above is related to your issue.

teukolsky commented:
Hi Geoffrey,

I think I was a little glib in saying what would happen if you swap 
interior and exterior quantities. zero-speed modes can be a little 
different. For example, for the scalar wave in curved spacetime, the 
component of the upwind flux corresponding to \psi is

- \betan \psi^-   (1)

Here \betan =\beta^a n_a is the normal component of the shift, which 
we've assumed is negative. So -\betan is positive, and the flux uses the 
interior value of \psi which is being advected toward the exterior.

If we interchange interior and exterior, the above expression becomes

+ \betan \psi^+   (2)

The sign change is because the unit normal changes sign when we change 
which side of the interface we're on. Note that this expression still 
describes \psi being advected in the same way from the left subdomain to 
the right subdomain. In other words, the conservation law is preserved: 
whatever leaves one subdomain enters the neighbor. Algebraically, 
however, expressions (1) and (2) are not obviously antisymmetric in the 
way some other terms in the flux are.

Best,

Saul

On 2/7/19 7:33 AM, Geoffrey Lovelace wrote:
>
> After discussing with @teukolsky <https://github.com/teukolsky>, I 
> modified this pull request to /not/ average when computing the upwind 
> flux. The *hpp file documents exactly what the current version does. 
> Briefly, here's what it does:
>
>  1. Multiply the positive char speeds (computed using only interior
>     quantities) by the corresponding interior char fields
>  2. Multiply the negative char speeds (computed using only interior
>     quantities) by the corresponding exterior char fields
>  3. Add the result from steps one and 2
>  4. Transform the result from step 3 to evolved variables, using
>     gamma2, the normal from the interior, and the result of step 3.
>
> I think I still don't have the algorithm right, though. I would have 
> expected I'd get the same result after step 4 (up to a sign) if I 
> repeat the steps swapping interior <-> exterior quantities, but I 
> don't. (Should I?)
>
> @teukolsky <https://github.com/teukolsky> @fmahebert 
> <https://github.com/fmahebert> do you have any ideas of how I can fix 
> my algorithm?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sxs-collaboration/spectre/pull/1322#issuecomment-461471954>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AeYvpzWQCz-tj2gaprIku7Oe-cqgCrlqks5vLEdEgaJpZM4amHc->.
>

wthrowe commented:
@geoffrey4444 I'm trying to understand the math here, so I'm just looking at the documentation.

One thing I am still confused by here is that you presumably have two \Lambda matrices, one from the inside and one from the outside, and there should be some indication of which one is used.  The other thing is that I am still uncertain about which quantities it makes sense to discuss the flux of.  From Saul's comment I'm guessing this is the evolved variables, but I am not sure of this.  This confusion may just reflect my relative unfamiliarity with this sort of system.  (I am also curious about the form of the S operator, but that I assume is in the reference and probably doesn't need to be reproduced here unless the details end up mattering.  I'll look it up for myself later.)

(I am assuming `\Lambda^{-1}` is a typo for `\Lambda^-`, and you have some typesetting errors that are making the math not render correctly.)

geoffrey4444 commented:
@wthrowe: thanks! You’re right about the typo...I’ll fix when I can. Lambda^+ and Lambda^- are diagonal matrices whose entries are the positive or negative char speeds, respectively. Currently the code gets both from the interior. The only exterior quantities used are the char fields. I share your unfamiliarity with this kind of system...I’ve mostly been going by Saul’s DG paper, but I’d welcome any pointers to other things I can read up on that might help!

nilsvu commented:
@geoffrey4444 This being the last remaining component for a GH Kerr system, I think it would help to get the executable working alongside the flux. Then it‘s clear how the flux fits into the system.

One reason I am confused is that the choice of Tags UMinus and UPlus clashes with the notation for the fields values on either side of an element interface (as far as I understand right now, please correct me if I am getting something wrong). I wrote more details in [my comment above](
https://github.com/sxs-collaboration/spectre/pull/1322#discussion_r254741224).

hrueter commented on docs/References.bib:
```diff
@@ -147,7 +148,7 @@ @book{Hartle2003gravity
   year       = "2003",
   publisher  = "Addison-Wesley",
-  url        = "https://www.pearson.com/us/higher-education/program/Hartle-Gravity-An-Introduction-to-Einstein-s-General-Relativity/PGM298455.html",
+  url        = "https://bit.ly/2MTD1Vt",
```
Why did you change this?

There is really no need to use a link shortener. If the service stops existing the link will be broken.

geoffrey4444 replied:
Because the line was longer than 80 characters, so I couldn't commit it otherwise

nilsvu replied:
Thought we had a check that allows committing if it detects the line is a link

geoffrey4444 replied:
If we do, it's not currently working. I just checked out the latest develop branch, then made a one character change on a different line in References.bib:

~~~~
$ git diff
diff --git a/docs/References.bib b/docs/References.bib
index cc4175e1..59915355 100644
--- a/docs/References.bib
+++ b/docs/References.bib
@@ -160,7 +160,7 @@
 }
 
   @book{Hartle2003gravity,
-  author     = "Hartle, J.B.",
+  author     = "Hartle, J.~B.",
   title      = "Gravity: An Introduction to {Einstein's} General Relativity",
   isbn       = "9780805386622",
   lccn       = "2002151118",
~~~~

Then I tried to commit and got this error:

~~~~
$ git add docs/References.bib
$ git commit -m "Test change References.bib"
Found lines over 80 characters:
docs/References.bib:100:                        and Duez, Matthew D. and Haas, Roland and Kidder, Lawrence
169:  url        = "https://www.pearson.com/us/higher-education/program/Hartle-Gravity-An-Introduction-to-Einstein-s-General-Relativity/PGM298455.html",
242:  title    = "The analytic theory of fluid disks orbiting the {Kerr} black hole",
~~~~

Am I doing something wrong? Or should I file an issue?

hrueter replied:
URLs should be filtered out from the 80 characters test.

/tools/FileTestDefs.sh states 
>     # This doesn't filter out URLs, but I can't think of a way to do
>     # that without breaking the highlighting.  They only get printed
>     # if there's another problem in the file.

I think the URL exceeding the 80 characters limit triggered because the line with 
`The analytic theory of fluid disks orbiting the {Kerr} black hole`
had 82 characters. Since you fixed that in this PR, it should be possible to commit with the long URL.

geoffrey4444 commented:
I updated this PR with the latest upwind flux from the branch where I've been evolving KerrSchild. I just noticed I missed one comment to remove a blank line, but otherwise, could someone please take another look at this?

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -1,16 +1,44 @@
+    const Scalar<DataVector>& char_speed_ext) noexcept {
+  typename FieldTag::type weighted_char_field = char_field_int;
+  DataVector char_speed_avg = get(char_speed_int);
+  char_speed_avg -= 0.5 * (get(char_speed_int) - get(char_speed_ext));
```
combine line 27 and 28 in one expression.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -1,16 +1,44 @@
+  DataVector char_speed_avg = get(char_speed_int);
+  char_speed_avg -= 0.5 * (get(char_speed_int) - get(char_speed_ext));
+
+  auto weighted_char_field_it = weighted_char_field.begin();
```
why are you initializing this outside of the for loop?

geoffrey4444 replied:
No good reason...just copied code from somewhere

geoffrey4444 replied:
Turns out there was a reason...I don't know how to get the code to compile if I initialize them all inside the for loop, because the compiler has trouble deciphering what auto means when I try that. If someone else has a solution, I'm happy to make the change, but I'm not sure it's worth it

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -1,16 +1,44 @@
+  auto weighted_char_field_it = weighted_char_field.begin();
+  for (auto int_it = char_field_int.begin(), ext_it = char_field_ext.begin();
+       int_it != char_field_int.end();
+       (void)++int_it, (void)++ext_it, (void)++weighted_char_field_it) {
```
remove the void casts.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -1,16 +1,44 @@
+  char_speed_avg -= 0.5 * (get(char_speed_int) - get(char_speed_ext));
+
+  auto weighted_char_field_it = weighted_char_field.begin();
+  for (auto int_it = char_field_int.begin(), ext_it = char_field_ext.begin();
```
Move line 26 to the line directly above the for loop.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+                  Frame::Inertial>& /*interface_unit_normal_ext*/) const
+    noexcept {
+  Scalar<DataVector> gamma1_avg = gamma1_int;
+  get(gamma1_avg) -= 0.5 * (get(gamma1_int) - get(gamma1_ext));
```
this is confusing. Combine 407 and 408 in one line.
```cpp
const Scalar<DataVector> gamma1_avg{ 0.5 * (get(gamma1_int) + get(gamma1_ext)) }
```

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  Scalar<DataVector> gamma1_avg = gamma1_int;
+  get(gamma1_avg) -= 0.5 * (get(gamma1_int) - get(gamma1_ext));
+  Scalar<DataVector> gamma2_avg = gamma2_int;
+  get(gamma2_avg) -= 0.5 * (get(gamma2_int) - get(gamma2_ext));
```
same here.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  const auto char_fields_int =
+      CharacteristicFieldsCompute<Dim, Frame::Inertial>::function(
+          gamma2_avg, inverse_spatial_metric_int, spacetime_metric_int, pi_int,
+          phi_int, interface_unit_normal_int);
```
Use the `compute_characteristic_fields`  function here.
to do this add the return by value version of the `compute_characteristic_fields`  function and remove the `compute_` prefix. (See issue #1480  )

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  const auto u_minus_int =
+      get<Tags::UMinus<Dim, Frame::Inertial>>(char_fields_int);
+
+  const Scalar<DataVector> char_speed_u_psi_int{char_speeds_int[0]};
```
#1481

geoffrey4444 replied:
Let's deal with #1481 in a future PR.

nilsvu replied:
yes of course!

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+      char_fields_int, char_speeds_int, char_fields_ext, char_speeds_ext);
+
+  const auto weighted_evolved_fields =
+      EvolvedFieldsFromCharacteristicFieldsCompute<Dim, Frame::Inertial>::
```
compute_evolved_fields_from_characteristic_fields

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+        char_fields_ext,
+    const typename Tags::CharacteristicSpeeds<Dim, Frame::Inertial>::type&
+        char_speeds_ext) const noexcept {
+  const auto u_psi_int = get<Tags::UPsi<Dim, Frame::Inertial>>(char_fields_int);
```
const reference

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+
+  const size_t n_pts = u_plus_int.begin()->size();
+  typename Tags::CharacteristicFields<Dim, Frame::Inertial>::type
+      weighted_char_fields{n_pts, 0.0};
```
make_with_value
remove `n_pts`

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +335,175 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  const auto u_minus_int =
+      get<Tags::UMinus<Dim, Frame::Inertial>>(char_fields_int);
+
+  const Scalar<DataVector> char_speed_u_psi_int{char_speeds_int[0]};
```
Don't copy into a new scalar, instead pass around as `const Datavector&`

hrueter reviewed (changes requested):
Group review by @nilsleiffischer (main), @hrueter 

This is only a partial review.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:
```diff
@@ -71,16 +85,10 @@ struct CharacteristicSpeedsCompute : Tags::CharacteristicSpeeds<Dim, Frame>,
   static typename Tags::CharacteristicSpeeds<Dim, Frame>::type function(
```
Instead of defining `function` try to assign the `characteristic_speeds` function. We are not sure what the syntax is to select the correct overload. In this case also remove the instantiation from the cpp.

If this does not work, move the definition into the .cpp and keep the instantiation.

geoffrey4444 replied:
Fair comment, but I think that is outside the scope of this PR...maybe it should be an issue?

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -447,45 +446,47 @@ void UpwindFlux<Dim>::operator()(
-  const Scalar<DataVector> char_speed_u_zero_int{char_speeds_int[1]};
-  const Scalar<DataVector> char_speed_u_plus_int{char_speeds_int[2]};
-  const Scalar<DataVector> char_speed_u_minus_int{char_speeds_int[3]};
+  const DataVector& char_speed_u_psi_int{char_speeds_int[0]};
```
We don't know what the behaviour for this constructor is and think that instead the assignment operator should be used.

nilsdeppe replied:
They're both the same. The assignment operator is never used, it's always the constructor, even if you type `=`. You can't assign to something that doesn't exist, but you can construct it.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp:
```diff
@@ -59,6 +59,20 @@ namespace GeneralizedHarmonic {
  * surface.
  */
 template <size_t Dim, typename Frame>
+typename Tags::CharacteristicSpeeds<Dim, Frame>::type characteristic_speeds(
```
The changes in `src/Evolution/Systems/GeneralizedHarmonic` should be in a seperate commit.

hrueter reviewed (commented):
group review by @nilsleiffischer (main), @hrueter

osheamonn commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -307,6 +333,178 @@ void ComputeNormalDotFluxes<Dim>::apply(
+      phi_int, interface_unit_normal_int);
+  const auto& char_speeds_int = characteristic_speeds(
+      gamma1_avg, lapse_int, shift_int, interface_unit_normal_int);
+  const auto& char_fields_ext = characteristic_fields(
```
Shouldn't the characteristics on the exterior side be calculated using the exterior unit normal?

geoffrey4444 replied:
No. That turns out to be unstable. The char fields on an element are always defined in terms of the normal of that element. The upwind flux only says that when you compute incoming char fields, use the evolved fields from the neighbor; for all other char fields, use quantities in the element.

wthrowe replied:
Won't that violate the requirement that the two elements calculate the same numerical flux?

geoffrey4444 replied:
I don’t think so. Empirically, as codes here is stable, but using the other normal is not. Physically, I think you’re doing the same thing on both sides: use neighbor incoming char fields, self outgoing char fields.

wthrowe replied:
I'm asking about correctness, not stability.

I was going to try to check this by modifying your test, but it looks like the test never actually calls anything except the internal `weight_char_fields` function.  Please modify your test so that it tests the public interface of your class.

osheamonn commented on src/Domain/FaceNormal.hpp:
```diff
@@ -76,6 +80,53 @@ struct UnnormalizedFaceNormal : db::ComputeTag {
+};
+
+template <size_t Dim, typename Frame>
+struct UnitFaceNormalCompute : UnitFaceNormal<Dim, Frame>, db::ComputeTag {
```
The tag `Tags::Normalized<Tags::UnnormalizedFaceNormal<...>>` is added to the Databox on each interface and external face, and  represents the unit normal one-form. Does this tag do anything more or different to that? If it doesn't I'd say to remove it.

nilsvu replied:
@geoffrey4444 added this new compute tag, that derives from a simple tag, to have a simple tag for the unit face normal that can be used for packaging in the flux. However, now this is duplicate code, since the `Normalized` tag has the same implementation. I think we have the following options here:
- Replace the implementation in this new compute tag by an aliases to `Tags::Normalized<Tags::UnnormalizedFaceNormal>::function` and so on. Then we still have two compute tags that do the same, but at least no code duplication. Tests are needed to test the new simple tag, compute tag, and equivalence to the old compute tag. Occurrences of `Tags::Normalized<Tags::UnnormalizedFaceNormal>` should probably be updated, but I don't like that there is an ambiguity in which compute tag to use, and the old one doesn't allow retrieval by the base tag.
- Just drop the additions you made to `FaceNormal.hpp` and add a private simple tag `UnitFaceNormal` to the flux class. That's what I initially suggested, is what is done in the other fluxes, and is really all you need here to package data for the flux. Changes to `FaceNormal.hpp` can then come in a separate PR, if we decide we want them.

wthrowe replied:
Remove the changes to FaceNormal.hpp and add a base tag for `Tags::Normalized`.

geoffrey4444 replied:
OK, I will stop using "UnitFaceNormal" etc. in my next update

osheamonn reviewed (changes requested):
This looks good overall. :+1: I especially like the tests you've written. Although I think there is too much done in one commit here. I think it should be split into possibly 3 commits
- Add UnitFaceNormalVector (And a test?)
- Remove 'compute_' prefix 
- Add Upwind.

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +144,167 @@ struct ComputeNormalDotFluxes {
+ * Here, \f$S^{-1}u^-\f$ represents the
+ * characteristic variables at the element interface, computed using
+ * evolved variables from the element interior;
+ * \f$S^{-1}u^+\f$ represents the characteristic variables at the element
```
I think it's quite confusing that `-` and `+` superscripts have three meanings, as far as I understand: Either "interior/exterior", "negative/positive" for `\Lambda`, or labels for particular characteristic fields. Please make this unambiguous in this documentation, e.g. by ditching the `-` and `+` entirely, to avoid confusion, and using `int`/`ext`, `neg`/`pos` and `minus`/`plus` for the three cases, respectively.

geoffrey4444 replied:
There are really two cases: int/ext for the fields and neg/pos for the sign of the speeds. I will use these instead of +/- in my next push to this PR, to avoid confusion with the individual fields U^+ and U^-.

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +144,167 @@ struct ComputeNormalDotFluxes {
+  // (if the char speed is outgoing) or the product of the exterior char field
+  // and its char speed (if the char speed is incoming).
+  db::item_type<Tags::CharacteristicFields<Dim, Frame::Inertial>>
+  weight_char_fields(
```
should this be private?

geoffrey4444 replied:
Yes, no one else should need this function

geoffrey4444 replied:
On second thoughts, doesn't this need to be public as long as my test calls it directly?

nilsvu commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,226 @@
+  // If all the char speeds are +1, the weighted fields should just
+  // be the interior fields
+  const auto weighted_char_fields_one = flux_computer.weight_char_fields(
+      char_fields_int, char_speeds_one, char_fields_ext, char_speeds_one);
```
I don't think you need the char fields to be computed from an actual solution for these tests, and could instead just generate random data for these, is that correct? Please try to remove all of the grid setup and analytic solution code, which is tested elsewhere and doesn't need to be tested here but makes up the majority of code in this test, and just `make_with_random_values` the `char_fields_int` and `char_fields_ext`.

nilsvu reviewed (changes requested):
Since I'll be on vacation until May 11th, @hrueter will replace me as a reviewer on this PR until I'm back.

nilsvu commented:
Oh and please squash the one fixup @geoffrey4444

geoffrey4444 committed:
Remove compute from GH characteristic function names

geoffrey4444 committed:
Add upwind flux

geoffrey4444 commented:
Responded to all comments except I didn't rewrite the test to use random values. I'll try that next

geoffrey4444 commented:
@wthrowe I think what I’m doing here is correct, but I’m not sure how to show that beyond testing stability. What would you like to do? I’m also not sure how to change the test as you request, because I don’t know any analytic solution to test against in that case. If you can help me with making a better test or with a way to verify correctness of the implementation, I welcome any ideas you might have!

geoffrey4444 commented:
I thought a bit more, @wthrowe, about your question. I do think using the same normal to compute the char fields gives the same flux on both sides of the interface. On both sides, use the exterior evolved fields to get the incoming char fields, and use the interior to get the outgoing fields, but what you mean by incoming and outgoing depends on the element’s unit normal, not the unit normal of the neighbor. But if there’s something wrong with this argument, I’m happy to discuss here or in a hangout

wthrowe commented:
For a test, you can at least check the general numerical flux consistency properties: `G(-F, -n, u; F, n, u) = F` and `G(F+, n+, u+; F-, n-, u-) = -G(F-, -n-, u-; F+, -n+, u+)` for any values of the arguments (might have some restrictions like `n` being normalized).  (I'm not entirely sure I got all the signs right there.)

geoffrey4444 commented:
@wthrowe I've added the first of your proposed consistency tests to my Schwarzschild test: put the same fields and equal-and-opposite normals into the upwind flux, and check that the numerical flux returned is the same as the flux. This test passes. @wthrowe I wonder if you could take a look when you get a chance at what I have so far?

It's not clear to me whether @wthrowe's second relation should be true in this case. I'm discussing with @teukolsky. Might you be free to discuss this as well at some point, @wthrowe? 

@nilsleiffischer I added a version of my previous test using random variables. But I don't think I can do these consistency tests using random variables, because that relies on spacetime_metric, pi, phi, lapse, shift, and inverse_spatial_metric all being consistent (so I'd have to numerically differentiate my randomly-generated spacetime_metric to get phi, and invert the spatial metric, for instance...this can especially cause trouble when randomly you get a length-0 norm for your normal vector, which causes FPEs).

nilsdeppe commented:
@geoffrey4444 I'm pretty sure @wthrowe's statement is a requirement for the scheme to be consistent. If the flux cannot be written in this way then I think the issue you have is that you don't conserve energy and an alternate method should be used. This might be why people use path-conservative methods for equations that cannot be written in flux conservative form. I'd like to be there when this is discussed.

geoffrey4444 commented:
We’ll definitely loop you in @nilsdeppe, and others interested in the discussion! I found @wthrowe’s first statement in Hesthaven’s book. Can anyone point me to the second statement in the literature?

nilsdeppe commented:
@geoffrey4444 thank you! 

The second statement is just conservation: what goes out of one element must go into the other. I'm not aware of any reference that writes it that explicitly. Partly the issue is if you don't have a dynamical metric then the relation is quite clear. It's trickier with a dynamical metric.

wthrowe commented:
> It's not clear to me whether @wthrowe's second relation should be true in this case. I'm discussing with @teukolsky. Might you be free to discuss this as well at some point, @wthrowe?

I am unavailable Monday and Tuesday of next week, but I should be available other days.

> 
> @nilsleiffischer I added a version of my previous test using random variables. But I don't think I can do these consistency tests using random variables, because that relies on spacetime_metric, pi, phi, lapse, shift, and inverse_spatial_metric all being consistent (so I'd have to numerically differentiate my randomly-generated spacetime_metric to get phi, and invert the spatial metric, for instance...this can especially cause trouble when randomly you get a length-0 norm for your normal vector, which causes FPEs).

The consistency checks should not need any relation between Psi, Pi, and Phi and should be tested on  data where those are inconsistent.  (The derivatives are expected to be off by truncation error because of the reduction to first order anyway.)  You will probably need to satisfy the algebraic relations of the 3+1 decomposition and the normal being normalized and such.

geoffrey4444 commented:
Let's discuss during Monday's spectre call, in the technical part of the discussion?

wthrowe commented:
I did just say I am unavailable Monday, so if you want me there that won't work.

geoffrey4444 commented:
I missed the "un" when I saw your post @wthrowe ... my bad! Let's make it during Wednesday's call then?

wthrowe commented:
Wednesday works for me.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -104,6 +104,7 @@ struct Normalized : db::ComputeTag {
     }
     return vector;
   }
+  using type = typename Tag::type;
```
This is not part of the interface, please don't add it.

geoffrey4444 replied:
Let’s chat to figure this out...it’s there because my GH executable doesn’t compile without it

geoffrey4444 replied:
To clarify: I added "using type = " here and below because I need to put `Tags::Normalized<Tags::UnnormalizedFaceNormal<>>` into a `Variables`, because I need the unit normal vector to be packaged with `package_data`. When I try to do this without these changes, I get compiler errors like the following:

~~~
In file included from /Users/geoffrey/Codes/spectre/spectre/src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:14:
/Users/geoffrey/Codes/spectre/spectre/src/DataStructures/Variables.hpp:111:44: error: no member named 'type' in 'Tags::Normalized<Tags::UnnormalizedFaceNormal<1, Frame::Inertial> >'
           cpp17::is_same_v<typename Tags::type::type,
                                     ~~~~~~^
/Users/geoffrey/Codes/spectre/spectre/src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:356:7: note: in instantiation of template class 'Variables<brigand::list<gr::Tags::SpacetimeMetric<1, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<1, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<1, Frame::Inertial>, gr::Tags::Lapse<DataVector>, gr::Tags::Shift<1, Frame::Inertial, DataVector>, gr::Tags::InverseSpatialMetric<1, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::ConstraintGamma1, GeneralizedHarmonic::Tags::ConstraintGamma2, Tags::Normalized<Tags::UnnormalizedFaceNormal<1, Frame::Inertial> > > >' requested here
      *packaged_data) = spacetime_metric;
      ^
/Users/geoffrey/Codes/spectre/spectre/src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:519:38: note: in instantiation of member function 'GeneralizedHarmonic::UpwindFlux<1>::package_data' requested here
template struct GeneralizedHarmonic::UpwindFlux<1>;
~~~

What is the proper way for me to resolve this compiler error, @nilsdeppe, if not what I've done here, so that I can put `Tags::Normalized<Tags::UnnormalizedFaceNormal<>>` into a `Variables`?

nilsdeppe replied:
I'd like to consult in person with @wthrowe about what the best solution is but this is behaving exactly as it should. You are trying to add a compute tag to a Variables which doesn't make sense since a Variables stores data, not lazily evaluated functions. What we probably want is to turn `Tags::Normalized` into a simple tag and add `Tags::NormalizedCompute` as a compute tag, doing so in such a way that we can use the base tag mechanism to retrieve `Tags::NormalizedCompute` using `Tags::Normalized`. Does that make sense?

geoffrey4444 replied:
I just realized this exact same issue is the subject of #1496

nilsdeppe commented on src/Domain/FaceNormal.hpp:
```diff
@@ -74,6 +74,7 @@ struct UnnormalizedFaceNormal : db::ComputeTag {
       tmpl::list<Mesh<VolumeDim - 1>, ElementMap<VolumeDim, Frame>,
                  Direction<VolumeDim>>;
   using volume_tags = tmpl::list<ElementMap<VolumeDim, Frame>>;
+  using type = tnsr::i<DataVector, VolumeDim, Frame>;
```
This is not part of the interface, please don't add it.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -1,16 +1,42 @@
 // Distributed under the MIT License.
 // See LICENSE.txt for details.
-
```
Please leave this blank line

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -317,3 +515,7 @@ template struct GeneralizedHarmonic::ComputeDuDt<3>;
 template struct GeneralizedHarmonic::ComputeNormalDotFluxes<1>;
 template struct GeneralizedHarmonic::ComputeNormalDotFluxes<2>;
 template struct GeneralizedHarmonic::ComputeNormalDotFluxes<3>;
+
```
Please switch this to use `GENERATE_INSTANTIATIONS` given how many things are now being instantiated.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +145,171 @@ struct ComputeNormalDotFluxes {
+ *
+ * The upwind flux in general is given by Eq. (6.3) of \cite Teukolsky2015ega :
+ * \f{eqnarray}{
+ * F^* = S \Lambda^{\rm pos} S^{-1} u^{\rm int}
```
I don't understand this equation. Where are the spatial indices? The numerical flux has an index. Comparing to Saul's paper this actually is `G`, not `F^*`.

geoffrey4444 replied:
This is exactly Eq. (6.3) of the paper I cite, except there was no name given to it. Looks like I should have just called this G here, and then everything's fine?

nilsdeppe replied:
I don't think it should be called G because it's not a flux but a different general boundary correction. It's also not actually a path-conservative boundary correction, which is often denoted by `D` in the literature, but I'd be fine glossing over that and just calling it `D`. The reason I don't like calling it `G` is that in its current form this does not satisfy the Rankine-Hugoniot conditions and so cannot lead to a well-balance scheme (this is just stating in math language what Will's concern is).

nilsdeppe reviewed (commented):
I've just figured out how to view the upwind flux in the framework of path conservative methods. I will write up an explanation of this in a tex document and upload that here so people can read it.

geoffrey4444 commented:
I now have both the random and analytic tests passing the first consistency check (recover the flux when identical inputs go into the numerical flux interior and exterior). I also derived this analytically for the upwind flux as presently coded.

Looking forward to seeing your note about the upwind flux @nils!

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp:
```diff
@@ -100,8 +99,7 @@ void compute_characteristic_fields(
 template <size_t Dim, typename Frame>
-typename Tags::CharacteristicFields<Dim, Frame>::type
-CharacteristicFieldsCompute<Dim, Frame>::function(
+typename Tags::CharacteristicFields<Dim, Frame>::type characteristic_fields(
```
This is an unsupported way of getting the type of a tag. Only `db::item_type` is supported. The more serious issue is that just using `db::item_type` here breaks the modular design of SpECTRE. The design is that `free functions -> DataBox`, while what you have here is `free functions <-> DataBox`. I think we should change these functions to be `free functions -> DataBox`. The free functions are the fundamental thing, not the DataBox tags.

geoffrey4444 replied:
I agree @nilsdeppe! But should this be in the same PR as the upwind flux? The only reason I'm touching this file at all is to rename the functions, so I can call them in the upwind flux using the correct naming convention. 

Whether we do it in a separate PR or not, to confirm, you're suggesting these free functions should explicitly specify their types, instead of using Tags::...::type, right? If so, to make sure I understand, should line 102 above be the following? If not, what should it be?

~~~
Variables<tmpl::list<Tags::UPsi<Dim, Frame>, Tags::UZero<Dim, Frame>,
                                    Tags::UPlus<Dim, Frame>, Tags::UMinus<Dim, Frame>>> characteristic_fields(
~~~

nilsdeppe replied:
Yes, I'm am saying we should use the types explicitly instead of their tags. I'm fine doing this as either a separate commit in this PR or a separate PR (submitted in a timely manner so that we don't have people write new code base of this code and then have a snowball effect to deal with :) ).

geoffrey4444 replied:
I’m happy to do this promptly, would prefer a separate PR that can go in quick. How do you avoid tags when using a Variables? Don’t you need tags for the keys?

nilsdeppe commented:
Here are my notes on comparing upwind and path-conservative methods, as well as having the boundary term satisfy a generalized Rankine-Hugoniot condition.

[UpwindFlux.pdf](https://github.com/sxs-collaboration/spectre/files/3184031/UpwindFlux.pdf)

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -8,9 +8,37 @@
+class Variables;
 
 // IWYU pragma: no_forward_declare Tensor
+namespace {
```
rename to `namespace GeneralizedHarmonic_detail {`

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -8,9 +8,37 @@
+
+  return weighted_char_field;
+}
+}  // namespace
```
`}  // namespace GeneralizedHarmonic_detail`

In your test you can then forward declare:
```cpp
namespace GeneralizedHarmonic_detail {
... weight_char_fields(...);
}
```

geoffrey4444 commented:
Restarted two Travis builds (gcc-6-debug and gcc-8-debug) that timed out

nilsdeppe commented:
@geoffrey4444 here is the tex file and an updated PDF of the document:


[UpwindFlux.pdf](https://github.com/sxs-collaboration/spectre/files/3187755/UpwindFlux.pdf)
[UpwindFlux.tex.txt](https://github.com/sxs-collaboration/spectre/files/3187756/UpwindFlux.tex.txt)
[UpwindFlux.bib.txt](https://github.com/sxs-collaboration/spectre/files/3187758/UpwindFlux.bib.txt)

GitHub is forcing me to upload them as `txt` files, so I just appended the `txt` extension to the tex and bib file.

kidder commented:
@wthrowe @osheamonn @nilsleiffischer @hrueter @nilsdeppe please look at

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,165 @@ struct ComputeNormalDotFluxes {
+ * the average value of the field
+ * \f$\gamma_2^{\rm avg} = (1/2)(\gamma_2^{\rm int} + \gamma_2^{\rm ext})\f$,
+ * where here \f$\gamma_2^{\rm int}\f$ is the value of \f$\gamma_2\f$
+ * in the interior and \f$\gamma_2^{\rm ext}\f$ is the vaule of
```
`value`

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,165 @@ struct ComputeNormalDotFluxes {
+ * \f$\gamma_2^{\rm avg} = (1/2)(\gamma_2^{\rm int} + \gamma_2^{\rm ext})\f$,
+ * where here \f$\gamma_2^{\rm int}\f$ is the value of \f$\gamma_2\f$
+ * in the interior and \f$\gamma_2^{\rm ext}\f$ is the vaule of
+ \f$\gamma_2\f$ in the exterior.
```
asterisk missing

hrueter reviewed (changes requested):
group review by @nilsleiffischer (main), @hrueter 

Two minor comments after a first look on the current code.

wthrowe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,165 @@ struct ComputeNormalDotFluxes {
+ * in the interior and \f$\gamma_2^{\rm ext}\f$ is the vaule of
+ \f$\gamma_2\f$ in the exterior.
+ *
+ * WARNING: as currently implemented, this flux does not satisfy the
```
`\warning`

wthrowe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,165 @@ struct ComputeNormalDotFluxes {
+  using options = tmpl::list<>;
+  static constexpr OptionString help = {
+      "Computes the generalized harmonic upwind flux. It requires no "
+      "options."};
```
The option parser will already print "Options: <No options>" when displaying the help, so you don't need to state it again.

wthrowe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,547 @@
+
+#include "ApparentHorizons/Strahlkorper.hpp"
+#include "ApparentHorizons/StrahlkorperGr.hpp"
+#include "ApparentHorizons/Tags.hpp"  // IWYU pragma: keep
```
Why is the numerical flux test using code from the horizon finder?

geoffrey4444 replied:
Because I'm testing the flux on a Strahlkorper, and I need to get the Strahlkorper's normal

kidder replied:
I don't see anything about a `Strahlkorper` so I am still confused by why you need things from ApparentHorizons

wthrowe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,547 @@
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.GeneralizedHarmonic.UpwindFlux",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/GeneralizedHarmonic/"};
```
You are not using python.

wthrowe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,547 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+#include "tests/Unit/TestingFramework.hpp"
```
Blank line above.

nilsvu commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,547 @@
+
+// Test GH upwind flux by comparing to Schwarzschild
+template <typename Solution>
+void test_upwind_flux_analytic(
```
Does this test anything that the random test can not? I can't see how using specific values for the spacetime quantities, instead of random values, adds anything to the test.

kidder commented:
@nilsleiffischer we often include a known analytic value in tests in addition to the random tests.  These are often quite useful in debugging if a change breaks the test, and also in catching bugs where the python and c++ both have the same mistake

nilsvu commented:
@kidder I understand that idea, but it’s not what the test named “analytic” in this PR actually does, as I understand it. Tests to analytic solutions should take a particular input and compare the output to actual numbers that we have computed externally. Alternatively/additionally we can use input so that the math reduces to a super simple expression that we have confidence the test implementation gets right, e.g. test each individual term in a big expression.
I think the test in this PR does neither, but tests the same things as the random test, but with specific values.

wthrowe commented:
I agree with @nilsleiffischer's assessment.  Good catch.

kidder commented:
ah okay, I didn't look at it closely, sorry for misunderstanding your comment @nilsleiffischer

kidder commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -305,13 +400,131 @@ void ComputeNormalDotFluxes<Dim>::apply(
+
+template <size_t Dim>
+void UpwindFlux<Dim>::operator()(
+    gsl::not_null<db::item_type<::Tags::NormalDotNumericalFlux<
```
please don't use `db::item_type` when the type is known...

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+
+  MAKE_GENERATOR(generator);
+  std::uniform_real_distribution<> dist(-1.0, 1.0);
+  std::uniform_real_distribution<> dist_positive(1.0, 2.0);
```
unused

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+
+  const auto& spatial_metric_int = gr::spatial_metric(spacetime_metric_int);
+  const auto& inverse_spatial_metric_int =
+      determinant_and_inverse(spatial_metric_int).second;
```
How does the const ref work here, since `determinant_and_inverse` returns by value?

geoffrey4444 replied:
c++ automatically extends the lifetime of the temporary object returned by determinant_and_inverse to the lifetime of the local const reference. If I understand this blog post correctly, the destructors are also called correctly when the const ref does go out of scope.

https://blog.galowicz.de/2016/03/23/const_reference_to_temporary_object/

So I think this is fine, but I would love to understand why not if I'm wrong!

nilsdeppe replied:
While it's technically not wrong in this case I view it as writing misleading code. The code as written does not correctly portray what you are intending, which clearly leads to confusion. Furthermore, there are cases (as pointed out in the article you linked) where very similar looking code leads to incorrect behavior. I believe if `.second` were a member function you'd have UB. I've generally been asking people to not do what you're doing here because of these subtleties.

geoffrey4444 replied:
OK, I removed the & here and below where I call det and inv

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+
+  // Get the characteristic fields and speeds
+  const auto char_fields_int = GeneralizedHarmonic::CharacteristicFieldsCompute<
+      spatial_dim, Frame::Inertial>::function(gamma_2,
```
Use the free function `characteristic_fields` instead of the compute item.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+  // Consistency checks
+  Variables<typename GeneralizedHarmonic::UpwindFlux<spatial_dim>::package_tags>
+  packaged_data_int(get(one).size(),
+                    std::numeric_limits<double>::signaling_NaN());
```
We prefer curly braces for initializations.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+                             gamma_2, unit_normal_one_form_ext);
+
+  const auto minus_unit_normal_one_form_int =
+      divide_by(unit_normal_one_form_int, get(minus_one));
```
This looks like an abuse of `divide_by` (which is not actually used in the code anywhere else).
Just loop over the components and use the `DataVector` operators.

geoffrey4444 replied:
I'm fine changing this, but why is this abuse? I would have thought this is fine, with the advantage of letting divide_by handle the looping to keep the code here simpler.

Also, why does spectre provide divide_by() if it's not used anywhere and should not be used in a case like this? When *should* divide_by be used?

nilsdeppe replied:
I wouldn't say it's so much of an abuse as it is just writing the code in a rather inefficient and clumsy manner... (though I can definitely see how this can be considered abuse :) ) You basically do a copy and create a datavector of all `-1` instead of doing a single for loop.

Looking at the implementation of `divide_by` it's not really clear to me what the intended use case is or why it exists at all. @osheamonn added it so hopefully he can answer these questions.

nilsvu replied:
We should probably add simple tensor arithmetic asap (as soon as someone finds the time and energy to do that) to support what a `Variables` can do already, i.e. addition and scalar multiplication at least...

geoffrey4444 replied:
In my latest commit, I removed the divide_by's, so this shouldn't hold up this PR. But I am curious what the result of the divide_by discussion is.

osheamonn replied:
`divide_by` was ported from bitbucket extremely early on in the porting process, when we were trying to port the small and easy things first. I presume that wherever it was used in bitbucket (probably to normalise unit vectors is my guess) didn't end up being ported, and that ended up being handled by the Interface items infrastructure which was added almost a year later. So yeah `divide_by` is probably not needed, but at the time it was added we thought it would be.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+  const auto& inverse_spatial_metric_int =
+      determinant_and_inverse(spatial_metric_int).second;
+  const DataVector one_over_one_form_magnitude_int =
+      get(magnitude(unnormalized_normal, inverse_spatial_metric_int));
```
This is not 1/|n| , but just the magnitude.

geoffrey4444 replied:
Good catch! I've fixed this so this is actually get(one) / get(magnitude(unnormalized_normal, inverse_spatial_metric_int)) and similarly for the ext case below

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+  const DataVector one_over_one_form_magnitude_int =
+      get(magnitude(unnormalized_normal, inverse_spatial_metric_int));
+  const auto unit_normal_one_form_int = StrahlkorperGr::unit_normal_one_form(
+      unnormalized_normal, one_over_one_form_magnitude_int);
```
Since `one_over_one_form_magnitude_int` is actually the magnitude it appears as if the tests are actually independent of the normalization of the one form.

geoffrey4444 replied:
That might be because I'm showing that normal dot numerical flux reduces to normal dot flux, so an overall normalization in the normal applies to both. Is there a specific change you would like in response to this comment?

nilsvu replied:
I suggest to produce random `unit_one_form`s (that perhaps don't even need to be normalized) instead of constructing the `Strahlkorper`. Then the only thing this test needs as input is a number of grid points (or a `DataVector used_for_size`).

geoffrey4444 replied:
I did that before seeing your comment :)

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+      normal_dot_flux_computer{};
+  auto psi_normal_dot_flux = make_with_value<
+      db::item_type<::Tags::NormalDotNumericalFlux<gr::Tags::SpacetimeMetric<
+          spatial_dim, Frame::Inertial, DataVector>>>>(one, 0.0);
```
initialize with NaN

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+  ::GeneralizedHarmonic::ComputeNormalDotFluxes<spatial_dim>
+      normal_dot_flux_computer{};
+  auto psi_normal_dot_flux = make_with_value<
+      db::item_type<::Tags::NormalDotNumericalFlux<gr::Tags::SpacetimeMetric<
```
use `::Tags::NormalDotFlux`

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+
+  CHECK_ITERABLE_APPROX(psi_normal_dot_numerical_flux, psi_normal_dot_flux);
+  CHECK_ITERABLE_APPROX(pi_normal_dot_numerical_flux, pi_normal_dot_flux);
+  CHECK_ITERABLE_APPROX(phi_normal_dot_numerical_flux, phi_normal_dot_flux);
```
Please add some structure to the code to explain what is being tested using scopes and `INFO`:
```cpp
{
     INFO(...);
}
```

Here you only test the consistency of the flux, i.e. F^i_num (u,u) = F^i(u) .
What is missing is conservation, i.e. the numerical flux is single valued when computed from either side. [[ F^i_num (u_int, n_int, u_ext, n_ext) ]]_i = n_i^int F^i_num(u_int, n_int, u_ext, n_ext) + n_i^ext F^i_num (u_ext, n_ext, u_int, n_int) = 0 .

geoffrey4444 replied:
The GH upwind flux as implemented does not satisfy conservation, as explained in its help text's warning. So I only test consistency here.

The plan (discussed earlier in this thread and on a call involving @nilsdeppe, @wthrowe, and others) is to first merge this flux as is. Then, I will look into alternatives, such as the one @nilsdeppe suggested, that do satisfy conservation. Since this flux is only intended for use on smooth solutions (i.e. vacuum), I suspect that it will work fine in general, though so far I have only tested it on KerrSchild.

nilsvu replied:
If you only test consistency, then you don't need the code that sets up the second randomly generated set of variables (you call them `_ext`). Please add a comment on what you are testing to the code nonetheless, and that you are not testing conservation.

geoffrey4444 replied:
I test consistency, but part of that test is that the weighting works properly, i.e. only returns the int or ext fields if the char speeds are + or -. So I do need the _ext quantities. But I do now comment that I'm testing consistency only.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,297 @@
+namespace {
+// Test GH upwind flux using random fields
+void test_upwind_flux_random(
+    const Strahlkorper<Frame::Inertial>& strahlkorper) noexcept {
```
This doesn't need to be a Strahlkörper. For your spatial normal one-forms you can choose random values. You only have to make sure, they are not zero.

hrueter reviewed (commented):
group review by @nilsleiffischer (main) and @hrueter

geoffrey4444 commented:
@nilsleiffischer @hrueter could you please take another look when you get a chance? I think I responded to all of your comments, but please let me know if I misunderstood or missed anything!

kidder commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,288 @@
+// IWYU pragma: no_forward_declare GeneralizedHarmonic::Tags::UZero
+// IWYU pragma: no_forward_declare GeneralizedHarmonic::Tags::UMinus
+// IWYU pragma: no_forward_declare GeneralizedHarmonic::Tags::UPlus
+// IWYU pragma: no_forward_declare StrahlkorperTags::CartesianCoords
```
remove these two pragmas

kidder commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,288 @@
+  const DataVector one_over_one_form_magnitude_ext =
+      get(one) /
+      get(magnitude(unit_normal_one_form_int, inverse_spatial_metric_ext));
+  const auto unit_normal_one_form_ext = StrahlkorperGr::unit_normal_one_form(
```
please remove this function from `StrahlkorperGr`, it is a generic function as far as I can tell; doesn't it exist elsewhere?

geoffrey4444 replied:
divide_by() does the same thing; I'll switch to that here and in a separate PR remove StrahlkorperGr::unit_normal_one_form.

geoffrey4444 replied:
PR #1571 removes StrahlkorperGr::unit_normal_one_form(), and I have removed it from this PR as well in my latest commit

nilsdeppe replied:
Please stop using `divide_by` and remove it from the code instead

geoffrey4444 replied:
So @nilsdeppe @kidder I should leave `StrahlkorperGr::unit_normal_one_form()` and still use that in the upwind flux? I'm confused because @kidder said there is a more generic function I should use, so remove the one from StrahlkorperGr, but that @nilsdeppe said no remove divide_by instead...do you two agree on what I should do here?

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,150 @@ struct ComputeNormalDotFluxes {
+      gr::Tags::Shift<Dim, Frame::Inertial, DataVector>,
+      gr::Tags::InverseSpatialMetric<Dim, Frame::Inertial, DataVector>,
+      Tags::ConstraintGamma1, Tags::ConstraintGamma2,
+      ::Tags::Normalized<::Tags::UnnormalizedFaceNormal<Dim, Frame::Inertial>>>;
```
In unrelated code I came across an issue that could affect your code here. There is code in `FluxCommunicationTypes.hpp` that tries to remove prefixes from these `package_tags`, which fails for your new `Tags::Normalized` (it's actually the `Variables` copy constructor that does that). Before merging this PR, please check if the code works with your GH executable. If you run into the same issue, I suggest to do what I have suggested before: make a new `FaceNormal` tag that is private to the class and use it to package the unit normal.

geoffrey4444 replied:
Thanks for the heads up! I just rebased on the latest develop, and this PR works with the GH executable, so I think we don't need to rework this before merging

geoffrey4444 commented:
OK, I think I resolved everyone's remaining comments...it's ready for another look

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -65,48 +62,41 @@ weight_char_fields(
-    const Strahlkorper<Frame::Inertial>& strahlkorper) noexcept {
+void test_upwind_flux_consistency_random() noexcept {
   const size_t spatial_dim = 3;
+  const size_t n_points = 5;
```
this appears to be unused.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+namespace {
+// Test GH upwind flux using random fields
+void test_upwind_flux_consistency_random() noexcept {
+  const size_t spatial_dim = 3;
```
`constexpr`

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+  const size_t n_points = 5;
+  const DataVector used_for_size{
+      {std::numeric_limits<double>::signaling_NaN(),
+       std::numeric_limits<double>::signaling_NaN()}};
```
use more points and the constructor `DataVector used_for_size{npoints, std::numeric_limits<double>::signaling_NaN()}`

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -208,8 +198,8 @@ void test_upwind_flux_random(
   Variables<typename GeneralizedHarmonic::UpwindFlux<spatial_dim>::package_tags>
-  packaged_data_int(get(one).size(),
-                    std::numeric_limits<double>::signaling_NaN());
+      packaged_data_int{get(one).size(),
```
use `used_for_size` and `make_with_value` here and below.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -197,7 +197,7 @@ struct ComputeNormalDotFluxes {
  * the exterior and vice versa. This condition is necessary for a
  * well-balanced scheme. Use this flux with caution; so far it has
- * been tested for evolving nonspinning Kerr-Schild in generalized harmonic.
+ * been tested for evolving aligned-spin Kerr-Schild in generalized harmonic.
```
what does "aligned" mean for a single black hole?

geoffrey4444 replied:
I'll just say "Kerr-Schild"...I meant spin on the z axis

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -246,15 +238,16 @@ void test_upwind_flux_random(
       db::item_type<::Tags::NormalDotNumericalFlux<gr::Tags::SpacetimeMetric<
-          spatial_dim, Frame::Inertial, DataVector>>>>(one, 0.0);
+          spatial_dim, Frame::Inertial, DataVector>>>>(
+      one, std::numeric_limits<double>::signaling_NaN());
```
`one` -> `use_for_size` here and below

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+      phi_int, unit_normal_one_form_int);
+  const auto char_fields_ext = GeneralizedHarmonic::characteristic_fields(
+      gamma_2, inverse_spatial_metric_ext, spacetime_metric_ext, pi_ext,
+      phi_ext, unit_normal_one_form_int);
```
`unit_normal_one_form_ext`

geoffrey4444 replied:
This is correct as coded. char_fields_ext are the char fields computed for the element using the fields from the neighbor but using the element's normal. cf. Equations.cpp:473

hrueter replied:
Maybe you can add a little comment, since other people might get confused as well.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  Variables<typename GeneralizedHarmonic::UpwindFlux<spatial_dim>::package_tags>
+  packaged_data_ext_opposite_normal(
```
`packaged_data_ext` and `packaged_data_ext_opposite_normal` are not needed for the consistency check.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.GeneralizedHarmonic.UpwindFlux",
+                  "[Unit][Evolution]") {
+  {
+    INFO("test consistency of the upwind flux")
```
Move the `INFO`s with the scoping into the test function and use it to separate the test of the weighting function from the consistency test.

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,284 @@
+                  "[Unit][Evolution]") {
+  {
+    INFO("test consistency of the upwind flux")
+    test_upwind_flux_consistency_random();
```
This function name is misleading since it tests more than consistency. Please remove `consistency` again.

hrueter reviewed (changes requested):
group review by @nilsleiffischer (main), @hrueter 

Please squash the previous fixups.

geoffrey4444 commented:
Put in latest requested fixups and squashed

hrueter commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp:
```diff
@@ -0,0 +1,266 @@
+                             shift_int, inverse_spatial_metric_int, gamma_1,
+                             gamma_2, minus_unit_normal_one_form_int);
+  tnsr::i<DataVector, spatial_dim, Frame::Inertial>
+      minus_unit_normal_one_form_ext = unit_normal_one_form_ext;
```
this is unused.
After removal `unit_normal_one_form_ext` should also become unused.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -127,4 +142,147 @@ struct ComputeNormalDotFluxes {
+ * which enforces that the flux leaving the interior is the flux entering
+ * the exterior and vice versa. This condition is necessary for a
+ * well-balanced scheme. Use this flux with caution; so far it has
+ * been tested for evolving Kerr-Schild in generalized harmonic.
```
in `generalized harmonic *gauge*`

geoffrey4444 replied:
No, in the generalized harmonic formulation of the Einstein equations.

geoffrey4444 replied:
I updated this line anyway, since now I've also tested for a perturbed black hole. I no longer mention generalized harmonic at all, for simplicity. (GH is the only formulation we have implemented for GR, and the only one we will implement any time soon).

hrueter reviewed (changes requested):
I took over for @nilsleiffischer 

Next time please only squash your old fixups and not the new ones.

hrueter commented:
The fixup looks good. You can squash.

geoffrey4444 commented:
Thanks very much for the quick reply, @hrueter! Squashed.

hrueter commented:
@osheamonn @wthrowe please take another look at this.

wthrowe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -305,13 +400,112 @@ void ComputeNormalDotFluxes<Dim>::apply(
-template struct GeneralizedHarmonic::ComputeNormalDotFluxes<1>;
-template struct GeneralizedHarmonic::ComputeNormalDotFluxes<2>;
-template struct GeneralizedHarmonic::ComputeNormalDotFluxes<3>;
+#undef GET_DIM
```
It's called `DIM` above.

wthrowe commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -305,13 +400,112 @@ void ComputeNormalDotFluxes<Dim>::apply(
-template struct GeneralizedHarmonic::ComputeNormalDotFluxes<2>;
-template struct GeneralizedHarmonic::ComputeNormalDotFluxes<3>;
+#undef GET_DIM
+#undef INSTANTIATION
```
It's called `INSTANTIATE` above.

wthrowe reviewed (changes requested):
It looks like there are some comments that are intended to be made into issues.  Make sure that gets done.

geoffrey4444 commented:
Thanks for the catches @wthrowe! I just pushed and squashed the trivial fixes for the two items you flagged. I also created issues that weren't already open for things raised in the above discussion.

nilsdeppe commented:
@kidder @nilsleiffischer please look at this. Thanks!
